### DISCUSSION TITLE
Support for custom classes for `raise`

### DISCUSSION THREAD

**Discussion by afuno** [sunny/actor#85] [Ideas]
@sunny What do you think about allowing (for example) the Rails side of the application to specify custom classes for `raise` that are used inside the gem?

For what? This idea came to my mind when I was trying to separate the use of Actor into services and forms. For example, there are two base classes:

```ruby
class ApplicationService
  include ServiceActor::Base
end
```

```ruby
class ApplicationForm
  include ServiceActor::Base
end
```

In this case, need to somehow catch errors from child classes. And now it will all be either `ServiceActor::ArgumentError` or `ServiceActor::Failure`.

Would like to prepare two custom classes on the Rails side and pass them inside the gem to `ApplicationService` and `ApplicationForm`.

**Comment by afuno** [comment]
An abstract example of what I meant:

```ruby
class ApplicationService
  include ServiceActor::Base

  argument_error_class(MyCustomErrorForService::ArgumentError)
  failure_class(MyCustomErrorForService::Failure)
end
```

```ruby
class ApplicationForm
  include ServiceActor::Base

  argument_error_class(MyCustomErrorForForm::ArgumentError)
  failure_class(MyCustomErrorForForm::Failure)
end
```

**Reply by afuno** [reply]
Draft: https://github.com/sunny/actor/pull/86

**Comment by afuno** [comment]
Prototype: https://github.com/afuno/actor/commit/7bd47235785c0f6aeb3c4c9398f8e1ca5b13c7c6

Example:

```ruby
module MyCustomErrorForService
  class ArgumentError < ServiceActor::ArgumentError; end

  class Failure < ServiceActor::Failure; end
end
```

```ruby
class ApplicationService
  include ServiceActor::Base

  define_argument_error_class(MyCustomErrorForService::ArgumentError)
  define_failure_class(MyCustomErrorForService::Failure)
end
```

```ruby
class PayService < ApplicationService
  input :order, type: String, allow_nil: false

  play :pay_order

  private

  def pay_order
    fail!("Can't pay for an order!")
  end
end
```

```ruby
PayService.call(order: nil)
```

```
MyCustomErrorForService::ArgumentError: The input "order" on PayService does not allow nil values
```

**Reply by afuno** [reply]
@sunny up 🙂

**Comment by sunny** [comment]
Added to v3.3.0 🎉
